Single character flags

neilpa opened the pull request:
Resolves #1. I plan to look at #3 as well but that depends on this change.

The one thing this doesn't handle perfectly is mixing short and long flags. It's biased to the last long form flag rather than the last of either. I expect this is probably fine though.

jspahrsummers commented on Commandant/ArgumentParser.swift:
```diff
@@ -40,6 +46,10 @@ extension RawArgument: Printable {
 			return "\"\(value)\""
+
+		case let .Flag(flags):
+			let combined = join("", map(flags) { String($0) })
```
I think you can replace this whole expression with `String(flags)`, because there's a constructor which accepts a `SequenceType`.

neilpa replied:
That's much better.

jspahrsummers commented on Commandant/Switch.swift:
```diff
@@ -0,0 +1,75 @@
+
+extension Switch: Printable {
+	public var description: String {
+		return key
```
Can we include the `flag` version of this switch here too?

jspahrsummers replied:
Maybe this isn't the right place, actually. Dunno.

neilpa replied:
I waffled on this one too which is why I left it off.

neilpa replied:
With my most recent changes to `Switch` this made sense to do now.

jspahrsummers commented on Commandant/Switch.swift:
```diff
@@ -0,0 +1,75 @@
+		return success(option.defaultValue)
+
+	case .Usage:
+		let description = option.defaultValue ? "--no-\(option.key)" : "--\(option.key)"
```
Can we include the single-character flag version here too? Something like `--verbose|v`?

neilpa replied:
Good catch

jspahrsummers commented:
This looks great, thanks! :sparkles:

> The one thing this doesn't handle perfectly is mixing short and long flags. It's biased to the last long form flag rather than the last of either. I expect this is probably fine though.

I want to agree so badly, but I feel like this kind of hampers composability. Many automated tools will append command line options blindly, expecting the last version of a given option to be the one that takes effect.

I'd like to be able to integrate with tools like that, which means we should ideally honor the last option specified, whether it was the long or short form.

neilpa commented:
> I want to agree so badly, but I feel like this kind of hampers composability. Many automated tools will append command line options blindly, expecting the last version of a given option to be the one that takes effect.

> I'd like to be able to integrate with tools like that, which means we should ideally honor the last option specified, whether it was the long or short form.

I started looking at #3 last night and as part of that already have a fix for this. I ended up wrapping the key names in an enum of string or character or both. That allows the argument parser to consume them together and return the last thing it finds.

neilpa commented:
@jspahrsummers I addressed your main concern about compose-ability but in doing so I changed the semantics of `Switch`.  Now it's a parameterless option that defaults to false. Once enabled it can't be disabled again. Boolean toggles will need to still use `Option<Bool>` without the single letter support.

This is less flexible than what I had before but it's consistent with the last writer wins behavior. Also it should be short-lived since I'm working on a mutually exclusive option set for #3. That'll allow long and/or short names for every choice.

jspahrsummers commented:
:+1: Thanks @neilpa!
